iothub_client/src/message_queue.c

```diff
@@ -386,32 +386,27 @@ static int move_messages_between_lists(SINGLYLINKEDLIST_HANDLE from_list, SINGLY
 
     while ((list_item = singlylinkedlist_get_head_item(from_list)) != NULL)
     {
+        MESSAGE_QUEUE_ITEM* mq_item = (MESSAGE_QUEUE_ITEM*)singlylinkedlist_item_get_value(list_item);
```

Comment:
>MESSAGE_QUEUE_ITEM* mq_item = (MESSAGE_QUEUE_ITEM*)singlylinkedlist_item_get_value(list_item); [](start = 8, length = 94)

if this is successful but the `singlylinkedlist_remove` fails, this seems like it is doing unnecessary work or worst case advancing this linked list forward one. i'm not sure how this works but might it be better to put this in an `else` after `singlylinkedlist_remove` and then checking if `mq_item is NULL before adding it.

Reply:
That was exactly the cause of the segfault. The original code tries to get the content of the list_item after remove it. :)

---
In reply to: [333103165](https://github.com/Azure/azure-iot-sdk-c/pull/1244#discussion_r333103165) [](ancestors = 333103165)

Reply:
i guess if it is just getting the value it won't advance anything but still seems like unnecessary work.

---
In reply to: [333103165](https://github.com/Azure/azure-iot-sdk-c/pull/1244#discussion_r333103165) [](ancestors = 333103165)

Reply:
oh interesting okay sounds good then

---
In reply to: [333107769](https://github.com/Azure/azure-iot-sdk-c/pull/1244#discussion_r333107769) [](ancestors = 333107769,333103165)